### DISCUSSION TITLE
Help hosting file ESP32 SDcard reader

### DISCUSSION THREAD

**Discussion by And8919** [micropython/micropython#10742] [General]
Hello, 
i'm designing a simple local web server to listen to music via html audio player.
 I using an ESP32 and a microSD card reader.  The local network connection works and I view my html page and the microSD reader writes and reads .txt files.  How can I access, via html code, an audio file in the sdcard?  Typing `source=SD/audio.mp3` in the html section, it finds nothing.  Thank you!

**Comment by bixb922** [comment]
What web server are you using? Is the route to the file defined in the web server? Does the request for the file get to the server? Is the file visible with os.stat("SD/audio.mp3")? Please include html statement, the <audio> tag uses src= and not source=

**Reply by And8919** [reply]
I use my ESP32 like server with my connection.
this is the code:
```

from time import sleep
import machine 
import sdcard
import uos

def web_page():
    file_path = '/SDCARD/1998.mp3'
    file = open(file_path, 'a')
  
    html = """
 <!DOCTYPE html>
<html>
    <style>
        body {background-color: DarkSlateGrey;}
        title {color: #DCDCDC;
               text-align: center;
        }
        h1    {color: #008B8B;
              text-align: center;
        } 
        p    {color: #008B8B;
              text-align: center;
        }
    </style>
        <title>ESP32 Web Server</title>
        <meta name="viewport" content="width=device-width, initial-scale=1">
        <body>
            <h1>Music page</h1>
                <p>Listen the music now!</p>
                    <p>
                        <audio controls>
                            <source src="/SDCARD/1998.mp3" type="audio/mpeg">
                                Your browser does not support the audio element.
                        </audio>
                    </p>
        </body>
</html> """
      
    return html





# Assign chip select (CS) pin (and start it high)
cs = machine.Pin(5, machine.Pin.OUT)

# Intialize SPI peripheral (start with 1 MHz)
spi = machine.SPI(1,
                  baudrate=1000000,
                  polarity=0,
                  phase=0,
                  bits=8,
                  firstbit=machine.SPI.MSB,
                  sck=machine.Pin(18),
                  mosi=machine.Pin(23),
                  miso=machine.Pin(19))

sd = sdcard.SDCard(spi, cs)
vfs = uos.VfsFat(sd)
uos.mount(vfs, "/SDCARD")


s = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
s.bind(('', 80))
s.listen(5)


while True:
  conn, addr = s.accept()
  print('Got a connection from %s' % str(addr))
  request = conn.recv(1024)
  request = str(request)
  response = web_page()
  conn.send('HTTP/1.1 200 OK\n')
  conn.send('Content-Type: text/html\n')
  conn.send('Connection: close\n\n')
  conn.sendall(response)
  conn.close()


```
I've tried 
```   
    file_path = '/SDCARD/1998.mp3'
    file = open(file_path, 'a')
```
but don't work.

**Comment by And8919** [comment]
I'm stuck. I can't find guides for my goal

**Reply by robert-hh** [reply]
@And8919 Did you try using the fill path name: /SD/audio.mp3, assuming that SDCard is mounted at /SD.

**Reply by And8919** [reply]
yes, the sd work.
If I use `w` or `r` a `.txt` file, it work.
But if I want insert in html a file from SD, don't work.
(I don't know how to do this)

**Comment by robert-hh** [comment]
Did you mount the sd card?

**Reply by And8919** [reply]
Yes

**Comment by bixb922** [comment]
In your code, there is no answer programmed when the browser requests your audio file. The html page will be processed by the browser, but then the browser will send a new request to your server for the audio mp3 file. Your code does not answer to that request.

The statement file = open(file_path, 'a') opens the audio file for append, please check that, I could not understand the purpose.

For a web server, Microdot is a good option, see https://github.com/miguelgrinberg/microdot

What you intend to do should be possible with Microdot.

**Reply by And8919** [reply]
Maybe microdot is the solution! Thanks

**Comment by And8919** [comment]
Ok, I'm back.

New code.

```
import network
from time import sleep
from microdot import Microdot
import machine
import sdcard
import uos

 
def do_connect():
    wlan = network.WLAN(network.STA_IF)
    wlan.active(True)
    if not wlan.isconnected():
        print('connecting to network...')
        wlan.connect('miowifi, 'miapassword')
        while not wlan.isconnected():
            pass
    print('network config:', wlan.ifconfig())
   
do_connect()
################

cs = machine.Pin(5, machine.Pin.OUT)
spi = machine.SPI(1,
                  baudrate=1000000,
                  polarity=0,
                  phase=0,
                  bits=8,
                  firstbit=machine.SPI.MSB,
                  sck=machine.Pin(18),
                  mosi=machine.Pin(23),
                  miso=machine.Pin(19))

sd = sdcard.SDCard(spi, cs)
vfs = uos.VfsFat(sd)
uos.mount(vfs, "/SDCARD")
f = open("/SDCARD/home.html", "r")  

################

app = Microdot()

@app.route('/')
def index(request):
    return f, {'Content-Type': 'text/html'}

app.run(port=80)
```

The file html in SDCARD work but the link to audio file don't.

Part of HTML code:
```
                        <audio controls>
                            <source src="/SDcard/1998.mp3" type="audio/mpeg">
                                Your browser does not support the audio element.
                        </audio>
```



Where I wrong?

**Reply by robert-hh** [reply]
You use ` <source src="/SDcard/1998.mp3" type="audio/mpeg">`, but the card was mounted to SDCARD.

**Reply by And8919** [reply]
Sorry, old code. On my sdcrd the code is use  is ```<source src="/SDCARD/1998.mp3" type="audio/mpeg">```

**Comment by And8919** [comment]
<img width="475" alt="img_1" src="https://user-images.githubusercontent.com/96819794/218676574-f5980a76-711d-4bb1-8814-167836c7371f.png">

microdot load correctly HTML but the link for media don't work (`<source src="/SDCARD/1998.mp3" type="audio/mpeg">`).
I don't know how fix it.
